File: volcano-compose/src/commonMain/kotlin/com/taewooyo/volcano/compose/HeatmapState.kt
/*
 * Copyright (C) 2026 taewooyo
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.taewooyo.volcano.compose

import androidx.compose.runtime.Composable
import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import com.taewooyo.volcano.heatmap.HeatmapNode
import com.taewooyo.volcano.heatmap.HeatmapPath
import com.taewooyo.volcano.heatmap.resolve

/** State holder for drill-down navigation in [Heatmap]. */
@Stable
public class HeatmapState internal constructor(
  private val root: HeatmapNode,
) {

  private var path: HeatmapPath by mutableStateOf(HeatmapPath())
  private var selectedLeaf: HeatmapNode? by mutableStateOf(null)

  public val visibleNode: HeatmapNode
    get() = root.resolve(path)

  /** The root-to-visible path, including the visible group. */
  public val breadcrumbs: List<HeatmapNode>
    get() {
      val nodes = mutableListOf(root)
      var current = root
      path.nodeIds.forEach { id ->
        val child = current.children.firstOrNull { it.id == id } ?: return nodes
        nodes += child
        current = child
      }
      return nodes
    }

  /** The leaf selected by the consumer, including its domain data, if any. */
  public val selectedNode: HeatmapNode?
    get() = selectedLeaf

  /** The id of [selectedNode], retained as a convenient lightweight selection value. */
  public val selectedId: String?
    get() = selectedLeaf?.id

  /** Whether a platform back action can return from the current group to its parent. */
  public val canNavigateUp: Boolean
    get() = path.nodeIds.isNotEmpty()

  /** Opens a direct child group of the currently visible node. */
  public fun drillDown(nodeId: String): Boolean {
    val target = visibleNode.children.firstOrNull { it.id == nodeId } ?: return false
    if (target.isLeaf) return false
    path = HeatmapPath(path.nodeIds + nodeId)
    return true
  }

  /** Navigates one level upward, returning whether the visible group changed. */
  public fun navigateUp(): Boolean {
    if (path.nodeIds.isEmpty()) return false
    path = HeatmapPath(path.nodeIds.dropLast(1))
    return true
  }

  /** Navigates to an item in [breadcrumbs]. */
  public fun navigateTo(nodeId: String): Boolean {
    val index = breadcrumbs.indexOfFirst { it.id == nodeId }
    if (index < 0) return false
    return navigateToBreadcrumb(index)
  }

  /**
   * Navigates to an item at [index] in [breadcrumbs].
   *
   * Prefer this over [navigateTo] for breadcrumb UI because node ids are only required to be
   * unique among siblings, not across an entire hierarchy.
   */
  public fun navigateToBreadcrumb(index: Int): Boolean {
    if (index !in breadcrumbs.indices || index == breadcrumbs.lastIndex) return false
    path = HeatmapPath(path.nodeIds.take(index))
    return true
  }

  /** Returns to the initial heatmap root. */
  public fun reset() {
    path = HeatmapPath()
  }

  /** Marks a leaf as selected. Selection does not change the visible navigation path. */
  public fun select(node: HeatmapNode) {
    if (node.isLeaf) selectedLeaf = node
  }

  /** Whether [node] is the exact leaf currently selected in this heatmap tree. */
  public fun isSelected(node: HeatmapNode): Boolean = selectedLeaf === node

  /** Clears the current leaf selection. */
  public fun clearSelection() {
    selectedLeaf = null
  }
}

/** Creates and remembers a drill-down state for [root]. */
@Composable
public fun rememberHeatmapState(root: HeatmapNode): HeatmapState = remember(root) {
  HeatmapState(root)
}

File: packages/volcano-react/src/HeatmapState.ts
import { useCallback, useEffect, useMemo, useState } from "react";
import type { HeatmapNode } from "./types";

/** React state holder for drill-down navigation and leaf selection. */
export interface HeatmapState {
  readonly visibleNode: HeatmapNode;
  readonly breadcrumbs: readonly HeatmapNode[];
  readonly selectedNode: HeatmapNode | null;
  readonly selectedId: string | null;
  readonly canNavigateUp: boolean;
  readonly drillDown: (nodeId: string) => boolean;
  readonly navigateUp: () => boolean;
  readonly navigateTo: (nodeId: string) => boolean;
  readonly navigateToBreadcrumb: (index: number) => boolean;
  readonly reset: () => void;
  readonly select: (node: HeatmapNode) => void;
  readonly isSelected: (node: HeatmapNode) => boolean;
  readonly clearSelection: () => void;
}

/** Creates state with the same navigation and selection operations as Compose [HeatmapState]. */
export function useHeatmapState(root: HeatmapNode): HeatmapState {
  const [path, setPath] = useState<readonly string[]>([]);
  const [selectedNode, setSelectedNode] = useState<HeatmapNode | null>(null);
  useEffect(() => {
    setPath([]);
    setSelectedNode(null);
  }, [root]);
  const breadcrumbs = useMemo(() => {
    const nodes = [root];
    let current = root;
    for (const id of path) {
      const child = current.children?.find((node) => node.id === id);
      if (!child) break;
      nodes.push(child);
      current = child;
    }
    return nodes;
  }, [root, path]);
  const visibleNode = breadcrumbs[breadcrumbs.length - 1];
  const drillDown = useCallback((nodeId: string) => {
    const target = visibleNode.children?.find((node) => node.id === nodeId);
    if (!target?.children?.length) return false;
    setPath((current) => [...current, nodeId]);
    return true;
  }, [visibleNode]);
  const navigateUp = useCallback(() => {
    if (path.length === 0) return false;
    setPath((current) => current.slice(0, -1));
    return true;
  }, [path.length]);
  const navigateToBreadcrumb = useCallback((index: number) => {
    if (index < 0 || index >= breadcrumbs.length - 1) return false;
    setPath(path.slice(0, index));
    return true;
  }, [breadcrumbs.length, path]);
  const navigateTo = useCallback((nodeId: string) => {
    const index = breadcrumbs.findIndex((node) => node.id === nodeId);
    return index >= 0 && navigateToBreadcrumb(index);
  }, [breadcrumbs, navigateToBreadcrumb]);
  const select = useCallback((node: HeatmapNode) => {
    if (!node.children?.length) setSelectedNode(node);
  }, []);
  const clearSelection = useCallback(() => setSelectedNode(null), []);

  return {
    visibleNode,
    breadcrumbs,
    selectedNode,
    selectedId: selectedNode?.id ?? null,
    canNavigateUp: path.length > 0,
    drillDown,
    navigateUp,
    navigateTo,
    navigateToBreadcrumb,
    reset: () => setPath([]),
    select,
    isSelected: (node) => selectedNode === node,
    clearSelection,
  };
}
